crates/hir-ty/src/infer/expr.rs

```diff
@@ -1681,9 +1681,8 @@ impl<'a> InferenceContext<'a> {
                 } else {
                     param_ty
                 };
-                if !coercion_target.is_unknown()
-                    && self.coerce(Some(arg), &ty, &coercion_target).is_err()
-                {
+                let coercion_target = self.insert_type_vars(coercion_target);
+                if self.coerce(Some(arg), &ty, &coercion_target).is_err() {
```

Comment:
Is this correct? I think we shouldn't emit type mismatch in this case:
```Rust
fn foo(x: Vec<Unknown>) {
}
foo(vec![1, 2, 3]);
```
So this is correct. But it feels like it was intentionally the way it was. Maybe it has performance consideration?

Reply:
Yeah I think this is correct thing to do. Can you comment that we insert type vars to avoid reporting that kind of type mismatches?

Also, can you add a test that *does* emit type mismatch when we fail to coerce? Something like

```rust
fn foo(x: Vec<Unknown>) {}
fn test() {
    foo(42);
      //^^ type mismatch here
}
```